### ISSUE TITLE
Question about CIoU implementation in this repo vs paper: w_temp variable?

### ISSUE THREAD

**Issue by glenn-jocher** [Zzh-tju/DIoU-SSD-pytorch#1]
Hi, I'm trying to implement CIoU loss in our https://github.com/ultralytics/yolov3 repo, and I was wondering what the `w_temp` variable does in the code. It is not mentioned in the paper, which hadiffes a slightly different implementation. Thanks!

https://github.com/Zzh-tju/DIoU-SSD-pytorch/blob/e9a86b9fe79ed6ea5e92ebecf52309d799ff9bab/utils/box/box_utils.py#L90-L94

**Comment by Zzh-tju** [comment]
When one do ∂(ar)/∂w1, ∂((w1 - w_temp) * h1)/∂w1 =h1; and when one do ∂(ar)/∂h1, ∂((w1 - w_temp) * h1)/∂h1 =w1-2w1=-w1. This correspond to our paper in ∂v w.r.t w and h.

**Comment by Zzh-tju** [comment]
When one do ∂(ar)/∂w1, ∂((w1 - w_temp) * h1)/∂w1 =h1; and when one do ∂(ar)/∂h1, ∂((w1 - w_temp) * h1)/∂h1 =w1-2w1=-w1. This correspond to our paper in ∂v w.r.t w and h.

**Comment by Zzh-tju** [comment]
When one do ∂(ar)/∂w1, ∂((w1 - w_temp) * h1)/∂w1 =h1; and when one do ∂(ar)/∂h1, ∂((w1 - w_temp) * h1)/∂h1 =w1-2w1=-w1. This corresponds to  the gradient of v w.r.t. w and h in our paper.

**Comment by glenn-jocher** [comment]
@Zzh-tju ok thanks, I will try to rerun our CIoU comparison with the exact implementation here.

**Comment by ghoshaw** [comment]
@Zzh-tju , How to implement CIOU loss in TF?
def CIOU (boxes1, boxes2)
    """
    :param boxes1: 
    :param boxes2:(xmin, ymin, xmax, ymax)
   
    """
    
    boxes1 = tf.concat([tf.minimum(boxes1[..., :2], boxes1[..., 2:]),
                        tf.maximum(boxes1[..., :2], boxes1[..., 2:])], axis=-1)
    boxes2 = tf.concat([tf.minimum(boxes2[..., :2], boxes2[..., 2:]),
                        tf.maximum(boxes2[..., :2], boxes2[..., 2:])], axis=-1)

    boxes1_area = (boxes1[..., 2] - boxes1[..., 0]) * (boxes1[..., 3] - boxes1[..., 1])
    boxes2_area = (boxes2[..., 2] - boxes2[..., 0]) * (boxes2[..., 3] - boxes2[..., 1])

    center_x1 = (bboxes1[..., 2] + bboxes1[..., 0]) / 2
    center_y1 = (bboxes1[..., 3] + bboxes1[..., 1]) / 2
    center_x2 = (bboxes2[..., 2] + bboxes2[..., 0]) / 2
    center_y2 = (bboxes2[..., 3] + bboxes2[..., 1]) / 2

    w1 = bboxes1[..., 2] - bboxes1[..., 0]
    h1 = bboxes1[..., 3] - bboxes1[..., 1]
    w2 = bboxes2[..., 2] - bboxes2[..., 0]
    h2 = bboxes2[..., 3] - bboxes2[..., 1]

  
    intersection_left_up = tf.maximum(boxes1[..., :2], boxes2[..., :2])
    intersection_right_down = tf.minimum(boxes1[..., 2:], boxes2[..., 2:])

  
    union_left_up = tf.minimum(boxes1[..., :2], boxes2[..., :2])
    union_right_down = tf.maximum(boxes1[..., 2:], boxes2[..., 2:])

    intersection = tf.maximum(intersection_right_down - intersection_left_up, 0.0)
    inter_area = intersection[..., 0] * intersection[..., 1]
    union_area = boxes1_area + boxes2_area - inter_area
    IOU = 1.0 * inter_area / (union_area+1e-8)#+1.0e-32

    intersection_diag = (center_x2 - center_x1)*(center_x2 - center_x1) + (center_y2 - center_y1)*(center_y2 - center_y1)
    union_xy = tf.clip_by_value(union_right_down - union_left_up, 0)

    union_diag = union_xy[..., 0]*union_xy[..., 0] + union_xy[..., 1]*union_xy[..., 1]

   
    arctan = tf.atan(w2/h2) - tf.atan(w1/h1)
    v = (4/(math.pi**2))*arctan*arctan
    alpha = v/(v+1-IOU)
    w_temp = 2*w1
    ar = (8/(math.pi**2))*arctan*((w1-w_temp)*h1)
    CIOU = IOU - (intersection_diag/(union_diag+1e-8)+alpha*ar)
    CIOU = tf.clip_by_value(CIOU, -1, 1)
    return CIOU


I am not sure how to implement the "with torch.no_grad():" part in TF or it does not matter?
